File: LeetCodeTasks/29_divide-two-integers/solutions/python3/solution.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        INT_MAX = 2**31 - 1
        INT_MIN = -2**31
        
        # Handling the signs and edge cases
        sign = -1 if (dividend < 0) ^ (divisor < 0) else 1
        
        # Use absolute values to simplify the division process
        dividend, divisor = abs(dividend), abs(divisor)
        
        # Division by repeated subtraction
        quotient = 0
        while dividend >= divisor:
            # This part of code tries to double the divisor each time to optimize the division process
            current_sum = divisor
            current_quotient = 1
            while dividend >= (current_sum << 1):
                current_sum <<= 1
                current_quotient <<= 1
            dividend -= current_sum
            quotient += current_quotient
        
        # Adjust the sign of quotient
        quotient *= sign
        
        # Handling overflow cases
        if quotient < INT_MIN:
            return INT_MIN
        elif quotient > INT_MAX:
            return INT_MAX
        else:
            return quotient

File: LeetCodeTasks/29_divide-two-integers/solutions/c/solution.c
#include <limits.h>

int divide(int dividend, int divisor) {
    // Special case for overflow
    if (dividend == INT_MIN && divisor == -1) {
        return INT_MAX;
    }
    
    // Determine the signs
    int sign = ((dividend < 0) ^ (divisor < 0)) ? -1 : 1;
    
    // Use long to handle extreme cases like INT_MIN
    long long ldividend = labs(dividend);
    long long ldivisor = labs(divisor);
    
    long long result = 0;
    
    // Perform division using bit manipulation and subtraction
    while (ldividend >= ldivisor) {
        long long temp = ldivisor, multiple = 1;
        while (ldividend >= (temp << 1)) {
            temp <<= 1;
            multiple <<= 1;
        }
        ldividend -= temp;
        result += multiple;
    }
    
    // Apply the sign
    return sign == 1 ? result : -result;
}
